### DISCUSSION TITLE
Assign color profile

### DISCUSSION THREAD

**Discussion by roelandschoukens** [ImageMagick/ImageMagick#2310] [Help] [answered]
Hi,

Is there a way to _assign_ an ICC or ICM color profile after loading an image? I.e. the equivalent of `-set colorspace`.

(as opposed to `-colorspace` and `-profile` which convert the pixel data)

**Comment by fmw42** [comment]
-profile does not change the pixel data. Use that to add a color profile.

**Comment by roelandschoukens** [comment]
OK, thanks. If so how do you do a conversion?

[The docs say this](https://imagemagick.org/script/command-line-options.php#profile):

> It is important to note that results may depend on whether or not the original image already has an included profile. Also, keep in mind that -profile is an "operator" (as opposed to a "setting") and therefore a conversion is made each time it is encountered, in order, in the command-line. For instance, in the following example, if the original image is CMYK with profile, a CMYK-CMYK-RGB conversion results.

Does it result in a conversion only if there is already a profile?

**Reply by fmw42** [reply]
That comment means that you need to consider whether you need one or two profiles when converting. If the input image has a profile, then you only need one to specify the output colorspace. If the input image does not have a color profile, then you need to provide a profile defining the colorspace for the input. So if you have a CMYK image and want to convert to sRGB, if it has a CMYK-like profile, you only need to add a sRGB profile in your command. If the CMYK image has no color profile, then you need to provide two profiles. The first must be a CMYK-like profile (not always having CMYK in the name) and the second profile would then be the sRGB profile. So two -profile commands. If you have an sRGB image and want to add a sRGB profile, you only need one profile command and I do not think it changes any data. The profile just tells the display or printer how to change the data when displaying or printing. But I will defer to the experts regarding changes in data values in the image.

**Comment by snibgo** [comment, marked as answer]
To _assign_ an ICC profile (which changes metadata but not pixel values), first remove any existing embedded profile (eg with "+profile"), then use "-profile" to assign the new profile. For example:

```
magick x.jpg +profile icc -profile mynewprofile.icc x2.jpg
```
If there was no embedded ICC profile, "+profile icc" will do nothing and is harmless.

**Reply by roelandschoukens** [reply]
Thanks!

Is there some built-in profile value that I can use to denote the default sRGB color space as the first _-profile_?

**Reply by fmw42** [reply]
Nothing built in that is automatic. You just have to provide -profile path_to/sRGB.icc  Most computer OS come with it. But you can get the profile from http://www.color.org/srgbprofiles.xalter

**Comment by snibgo** [comment]
As Fred says. IM does not contain any built-in ICC profiles, and you need to give an ICC filename to "-profile".

Your IM installation directory probably contains "sRGB.icc", which you can use.

When a file has no embedded ICC profile, most software will assume it is sRGB.

When IM writes a PNG file, if the image has an embedded ICC sRGB profile, IM will remove the profile and write a (short) sRGB chunk instead. In theory this does no damage, and saves space. But in practice sRGB ICC profiles vary, so I suggest using "-define png:preserve-iCCP". See https://imagemagick.org/script/defines.php

**Reply by Paul3367** [reply]
Hello snibgo,

I'm having troubles with JPG files that loose their embedded profiles. Would your suggestion _("-define png:preserve-iCCP")_ work with jpg files as well?

Thanks in advance

**Comment by snibgo** [comment]
> Would your suggestion ("-define png:preserve-iCCP") work with jpg files as well?

No, because the behaviour I described applies when writing PNG, not JPG.
